pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/OffHeapStringStore.java

```diff
@@ -0,0 +1,235 @@
+ *                  |        .....                   |
+ *                  |  start offset of string N      |
+ *                  |                                |
+ *                  |         UNUSED                 |
+ *                  |                                |
+ *                  | STRING N .....                 |
+ *                  |          .....                 |
+ *                  |          .....                 |
+ *                  | STRING N-1                     |
+ *                  |          .....                 |
+ *                  |          .....                 |
+ *                  | STRING 0 .....                 |
+ *                  |          .....                 |
+ *                  |          .....                 |
+ *                  |________________________________|
+ *
+ *
+ * We fill the buffer as follows:
+ * - The strings are added from the bottom, each new string appearing nearer to the top of the buffer, leaving no
+ *   room between them. Each string is stored as a sequence of 'char' elements. (In Java, each char element takes 2
+ *   bytes. A string has string.length() char elements in it).
+ *
+ * - The start offsets of the strings are added from the top. Each start offset is stored as an integer, taking 4 bytes.
+ *
+ * Each time we want to add a new string, we check if we have space to add the length of the string, and the string
+ * string itself. If we do, then we compute the start offset of the new string as:
+ *
+ *    new-start-offset = (start offset of prev string added) - (length of this string)
+ *
+ * The new start offset value is stored in the offset
+ *
+ *    buffer[numStringsSoFar * 4]
+ *
+ * and the string itself is stored starting at new-start-offset
+ *
+ */
+public class OffHeapStringStore implements Closeable {
+  private static class Buffer implements Closeable {
+    private static final int INT_SIZE = V1Constants.Numbers.INTEGER_SIZE;
+    private static final int CHAR_SIZE = Character.SIZE / 8;
+
+    final PinotDataBuffer _pinotDataBuffer;
+    final ByteBuffer _byteBuffer;
+    final int _startIndex;
+    final long _size;
+
+    int _numStrings = 0;
+    int _availEndOffset;  // Exclusive
+
+    private Buffer(long size, int startIndex) {
+      if (size > Integer.MAX_VALUE) {
+        size = Integer.MAX_VALUE;
+      }
+      _pinotDataBuffer = PinotDataBuffer.allocateDirect(size);
+      _pinotDataBuffer.order(ByteOrder.nativeOrder());
+      _byteBuffer = _pinotDataBuffer.toDirectByteBuffer(0, (int) size);
+      _startIndex = startIndex;
+      _availEndOffset = _byteBuffer.capacity();
+      _size = size;
+    }
+
+    private int add(String string) {
+      int startOffset = _availEndOffset - string.length() * CHAR_SIZE;
+      if (startOffset <= (_numStrings + 1) * INT_SIZE) {
+        // full
+        return -1;
+      }
+      for (int i = 0, j = startOffset; i < string.length(); i++, j = j + CHAR_SIZE) {
```

Comment:
I did a second implementation with byte (i.e. this class became a OffHeapByteStore instead of a string store). I re-factored the base class so that I convert incoming string into UTF-8 before doing any operation. Initial results were not nice, since running the unit tests took forever. I suspected gc, but could not follow up on things yesterday. I decided to run a benchmark, kicked one off last night, but it did not complete for some reason. Perhaps because the buffers were not being doubled each invocation. I iwll try to run the benchmark today to get some answer on this.